calico/felix/plugins/fiptgenerator.py

```diff
@@ -0,0 +1,731 @@
+        # * Start by marking all packets.
+        # * If we hit an allow rule, we'll return with the mark still set, to
+        #   indicate that we matched.
+        # * If we hit a deny rule, we'll drop the packet immediately.
+        # * If we reach the end of the chain, we'll un-mark the packet again to
+        #   indicate no match.
+
+        updates = {}
+        deps = {}
+
+        for direction in ("inbound", "outbound"):
+
+            chain_name = self._profile_to_chain_name(direction, profile_id)
+            rules_key = "%s_rules" % direction
+            rules = profile.get(rules_key, [])
+
+            fragments = [
+                '--append %s --jump MARK --set-mark 1' % chain_name
+            ]
+            for r in rules:
+                rule_version = r.get('ip_version')
+                if rule_version is None or rule_version == ip_version:
+                    fragments.extend(self._rule_to_iptables_fragments(
+                        chain_name,
+                        r,
+                        ip_version,
+                        tag_to_ipset,
+                        on_allow=on_allow,
+                        on_deny=on_deny))
+
+            # If we get to the end of the chain without a match, we remove the
+            # mark again to indicate that the packet wasn't matched.
+            fragments.append(
+                '--append %s '
+                '--match comment --comment '
+                '"No match, fall through to next profile" '
+                '--jump MARK --set-mark 0' % chain_name
+            )
+            updates[chain_name] = fragments
+
+        return updates, deps
+
+    def drop_rules(self, ip_version, chain_name, rule_spec, comment,
+                   ipt_action="--append"):
+        """
+        Return a list of iptables updates that can be applied to a chain to
+        drop packets that meet a given rule_spec.
+        :param ip_version.  Whether these are for the IPv4 or IPv6 iptables.
+        :param chain_name: the chain that the updates will be applied to
+        :param rule_spec: the rule spec (e.g. match criteria).   May be None
+        to drop all packets.
+        :param comment: any comment that should be associated with the
+        rule.  May be None to not include a comment.
+        :param ipt_action: the action that should be used to apply the rule
+        (e.g. --append or --insert)
+
+        :return list: a list of iptables fragments of the form
+        [ipt_action] [chain_name] [rule_spec] [action] [comment] e.g.
+        --append my_chain --match conntrack --ctstate INVALID --jump DROP
+        """
+        comment_str = None
+        if comment is not None:
+            comment = comment[:255]  # Limit imposed by iptables.
+            assert re.match(r'[\w: ]{,255}', comment), \
+                "Invalid comment %r" % comment
+            comment_str = '-m comment --comment "%s"' % comment
+
+        return [' '.join(filter(None, [
```

Comment:
`filter(None,`, I find the list comprehension easier to read (but might be personal pref)
```
parts = [p for p in [ipt_action, chain_name, rules_spec, '--jump DROP', comment_str] 
         if p is not None]
return ' '.join(parts)
```